Property accessors 6.2.0

jasongin opened the pull request:
See https://github.com/nodejs/abi-stable-node/issues/44

This change adds a new `napi_set_accessor` API which sets getter and setter accessors for a property on an object. In support of that, I refactored some of the callback-related code: the callback info opaque pointer type is now shared between function callbacks and accessor callbacks, so I renamed it from `napi_func_cb_info` to `napi_callback_info`, and renamed a few other things accordingly. In the internal implementation, there is a new `CallbackInfoWrapper` and subclasses that wrap the v8 `FunctionCallbackInfo` and `PropertyCallbackInfo` classes.

I assume we're not too concerned about breaking changes at this point, though of course any existing code using NAPI will need to be updated. I am updating the `abi-stable-node-addon-examples` repo for these changes, along with new examples for using property accessors.

jasongin committed:
napi: Property accessors and callback data

 - Add support for property descriptors including accessor callbacks
 - Add a data pointer for function and accessor callbacks
 - Add API for defining a property using a property descriptor
 - Update the constructor API to use property descriptors

jasongin commented:
Based on I've learned today I'm aware of at least a couple pieces missing from this PR:

1. `NAPI_GETTER` and `NAPI_SETTER` macros in `node_api_helpers.h`
2. Changes to `napi_create_constructor_for_wrap_with_methods` to also support accessors.

I'm working on these, but meanwhile I'd still appreciate feedback on the rest of the changes.

mhdawson reviewed (commented):
I'll need a bit more time to go through this in more detail (sounds like there are more changes to come anyway), but overall it looks reasonable to me.

One specific question sor far is if we can move some of the common functions into CallbackInfoWrapper instead of duplicating them in each type (for example: This(), Holder() , SetReturnValue, etc.)

mhdawson commented:
On the breaking changes front, agreed we still need to make breaking changes and as you say we need to then update our existing module ports and the add on examples so they continue to work.  

On that front I'm thinking we might want to set up some ci jobs, I'll think about that a bit more when I get back from vacation.

jasongin commented:
Regarding `CallbackInfoWrapper`, I did spend some time trying to come up with something less repetitive. It's tricky because v8's `FunctionCallbackInfo` and `PropertyCallbackInfo` have a lot of the same methods but don't have any inheritance relationship. I tried doing something with templates, but it got very complicated to the point that I thought it wasn't worth it.

jasongin commented:
I pushed another commit that includes the following changes: 
 * Changed `napi_method_descriptor` to `napi_property_descriptor` and added several fields. This is now a C++ equivalent to the property descriptor object passed to the JS `Object.defineProperty()`.
 * Updated `napi_create_constructor()` to take an array of property descriptors (instead of just the method descriptors), so it is able to define methods, accessors, and values for the class.
 * Added `napi_define_property()` that also takes a property descriptor and uses it to define a method, accessors, or value property on an object.
 * Refactored the internal callback wrapper helper classes and functions so they are cleaner and support passing a `void*` data pointer to any function or accessor callback.
 * Added `napi_cb_get_data()` to retrieve the callback data.
 * Added `NAPI_GETTER` and `NAPI_SETTER` macros in `node_api_helpers.h`. They're actually just the same as `NAPI_METHOD` because the args and return value are all handled via the callback info object in the same way. (Unlike nan where `NAN_SETTER` declares the value as a separate callback arg.)

I think it's ready for a full review now.

jasongin commented:
Oh, I still need to add/update tests covering these changes under `test/addons-abi`.

jasongin commented:
Pushed another commit with added/updated tests.

gabrielschulhof commented:
LGTM

jasongin commented:
Waiting for the [try-catch PR](https://github.com/nodejs/abi-stable-node/pull/49) to be merged, then I will rebase this PR on top of that.

jasongin commented:
I squashed and rebased on top of the latest changes. ~~but there are some test failures now. Please stand by.~~

jasongin commented:
OK I fixed the problems after rebasing. This is ready now. Waiting for signoff from @boingoing who said he would review this afternoon.

boingoing commented on src/node_jsvmapi.cc:
```diff
@@ -316,17 +522,47 @@ napi_value napi_create_constructor_for_wrap_with_methods(
+        v8::AccessControl::DEFAULT,
+        attributes);
+    }
+    else {
```
Just a nitpick, we could assert p->value here or raise an error condition if it's null. Fine to do this as part of the error handling pass later, though.

jasongin replied:
I was assuming we should preserve whatever is the existing behavior of PrototypeTemplate()->Set() when it is given a null value.

boingoing commented on src/node_jsvmapi.cc:
```diff
@@ -581,63 +864,51 @@ napi_value napi_get_true(napi_env e) {
+  return info->Holder();
+}
+
+void* napi_get_cb_data(napi_env e, napi_callback_info cbinfo) {
```
Handling of this data pointer is clean throughout. Good stuff.

boingoing commented:
This looks very good to me. Thanks a lot for doing the work, especially for updating the tests and adding the data pointer to the callback info.

mhdawson reviewed (commented):
A bit late but going through the changes look good to me. 

One last question.  Do you think we still need napi_set_property ? The places where it was replaced by napi_define_property seem more concise.

jasongin commented:
Yes, we still need `napi_set_property`, because sometimes you just want to assign a value to a property (which will invoke any existing set accessor if there is one) without redefining the property.
